source/OrderedTypes/ZornsLemma.lagda

```diff
@@ -0,0 +1,375 @@
+open import Ordinals.OrdinalOfOrdinals ua
+open import Ordinals.BuraliForti ua
+open import UF.ImageAndSurjection pt
+
+fe : FunExt
+fe = Univalence-gives-FunExt ua
+pe : Prop-Ext
+pe = Univalence-gives-Prop-Ext ua
+pe' : PropExt
+pe' 𝓤 = pe {𝓤}
+fe' : Fun-Ext
+fe' {𝓤} {𝓥} = fe 𝓤 𝓥
+
+open UF.Choice.choice-functions pt pe' fe
+open UF.Choice.Univalent-Choice fe pt
+open import Ordinals.Arithmetic fe
+open import OrderedTypes.Poset fe'
+open import Ordinals.WellOrderingTaboo fe' pe
+open PosetAxioms
+open ClassicalWellOrder pt
+open UF.Logic.AllCombinators pt fe'
+open UF.Choice.ExcludedMiddle pt fe
+open inhabited-subsets pt
+
+_⋘_ : X → X → (𝓤 ⊔ 𝓣) ̇
+a ⋘ b =  a ≪ b × (a ≠ b)
+
+is-chain : {𝓥 : Universe} → (Y : 𝓟 {𝓥} X ) →  (𝓤 ⊔ 𝓣 ⊔ 𝓥) ̇
+is-chain Y = (x y : X) → x ∈ Y → y ∈ Y → ∥ (x ≪ y + y ≪ x) ∥
+
+has-maximal-element-strong :  (𝓤 ⊔ 𝓣) ̇
+has-maximal-element-strong = Σ x ꞉ X , ((y : X) → x ≪ y → x ＝ y)
+
+all-chains-have-upper-bound : {𝓥 : Universe} → (𝓤 ⊔ 𝓣 ⊔ (𝓥 ⁺))  ̇
+all-chains-have-upper-bound {𝓥} =
+ (Y : 𝓟 {𝓥} X) → (is-chain Y) → Σ x ꞉ X , (∀ y → y ∈ Y → y ≪ x)
+
+\end{code}
+
+The following is not a direct formalization of any extant proof,
+but follows the same proof idea as the standard transfinite-induction-based
+proof of Zorn's lemma. The closest informal analog to the proof can be found
+here: https://math.stackexchange.com/a/2889205 .
+
+We do a proof by contradiction (which is why we have that double negation),
+where we assume that even though all chains have a least upper bound,
+the set at hand does not have a maximal element.
+These assumptions allow us to construct a function `f' from the ordinals to X,
+which is a set, which preserves order, is therefore left-cancellable, and again
+in turn an equivalence.
+`f' being order preserving relies on the fact that there is no maximal element,
+as otherwise `f' would (as defined) map many ordinals to the maximal element.
+An equivalence between the ordinals and a set is impossible,
+so there must (not-not) be a maximum.
+
+While this is the core of the proof logic, the final result to be used elsewhere
+is found below, and takes in more reasonable inputs.
+
+\begin{code}
+
+choice-function-gives-not-not-zorns-lemma :
+ Excluded-Middle →
+ poset-axioms (_≪_) →
+ all-chains-have-upper-bound  →
+ (Σ ε ꞉ (𝓟 {𝓣 ⊔ 𝓤} X → X) , ((A : 𝓟 X) → is-inhabited A → ε A ∈ A)) →
+ ¬¬ has-maximal-element-strong
+
+choice-function-gives-not-not-zorns-lemma
+ lem
+ (X-is-set , ≪-prop , ≪-refl , ≪-trans , ≪-antisym )
+ chains
+ (ε , ε-behaviour)
+ no-max =
+  absurd
+   where
+    dn : {𝓥 : Universe } → DNE 𝓥
+    dn {𝓥} = EM-gives-DNE lem
+
+    eq-is-¬¬-stable : {x : X} {y : X } →  ¬¬-stable (x ＝ y)
+    eq-is-¬¬-stable {x} {y} ¬¬x=y = dn (x ＝ y) (X-is-set) ¬¬x=y
+
+    ≪-is-¬¬-stable : {x : X} {y : X} → ¬¬-stable (x ≪ y)
+    ≪-is-¬¬-stable {x} {y} ¬¬ll = dn (x ≪ y) (≪-prop x y) ¬¬ll
+
+    ¬¬Σ→∃ : {𝓤 : Universe} {A : X → 𝓤  ̇} → ¬ ¬ (Σ x ꞉ X , A x ) → (∃ x ꞉ X , A x)
```

Comment:
Maybe this could go to `UF.ClassicalLogic` or somewhere else, given that it is a fact of general use when one is doing classical logic.